File: python/analyzer.py
import ast
import os
import time
from neo4j import GraphDatabase

# Node color mappings
NODE_COLORS = {
    'Root': 'orange',
    'Class': '#4287f5',
    'Function': '#42f54e',
    'Method': '#42f54e',
    'ExternalService': '#f5f542',
    'MongoCollection': '#f5a442',
}

def create_neo4j_driver():
    uri = os.getenv("NEO4J_URI")
    username = os.getenv("NEO4J_USER")
    password = os.getenv("NEO4J_PASSWORD")

    if not uri or not username or not password:
        raise RuntimeError("NEO4J_URI, NEO4J_USER, and NEO4J_PASSWORD environment variables are required.")

    driver = GraphDatabase.driver(uri, auth=(username, password))
    max_retries = 100
    delay = 5
    for attempt in range(max_retries):
        try:
            with driver.session() as session:
                session.run("RETURN 1")
            print("Connected to Neo4j")
            return driver
        except Exception as e:
            print(f"Attempt {attempt + 1}: Neo4j connection failed ({e}), retrying in {delay} seconds...")
            time.sleep(delay)
            delay *= 2
    raise Exception("Failed to connect to Neo4j after several attempts")

def cleanup_previous_run_data(session, root_name):
    session.run("MATCH (n) WHERE n.project = $project DETACH DELETE n", {"project": root_name})
    print("Cleaned up data from previous run")

def create_uniqueness_constraints(session):
    # Add a constraint for Root as well
    session.run("CREATE CONSTRAINT IF NOT EXISTS FOR (r:Root) REQUIRE r.id IS UNIQUE")
    session.run("CREATE CONSTRAINT IF NOT EXISTS FOR (c:Class) REQUIRE c.id IS UNIQUE")
    session.run("CREATE CONSTRAINT IF NOT EXISTS FOR (f:Function) REQUIRE f.id IS UNIQUE")
    session.run("CREATE CONSTRAINT IF NOT EXISTS FOR (m:Method) REQUIRE m.id IS UNIQUE")

def analyze_file(file_path, session, root_name, base_url):
    with open(file_path, 'r', encoding='utf-8') as file:
        code = file.read()
    project_root = '/app'

    def create_url(file_path, line_number):
        relative_path = os.path.relpath(file_path, project_root)
        return f"{base_url}/{relative_path}#{line_number}"

    try:
        tree = ast.parse(code, filename=file_path)
        AnalyzerVisitor(session, root_name, file_path, create_url).visit(tree)
    except Exception as e:
        print(f"Error parsing {file_path}: {e}")

class AnalyzerVisitor(ast.NodeVisitor):
    def __init__(self, session, root_name, file_path, create_url):
        self.session = session
        self.root_name = root_name
        self.file_path = file_path
        self.create_url = create_url
        self.current_class = None
        self.root_id = f"root:{self.root_name}"

    def merge_node(self, label, node_id, properties):
        # Merge a node with ON CREATE and ON MATCH to keep properties updated
        set_clauses = ", ".join([f"n.{key} = ${key}" for key in properties.keys()])
        query = (f"MERGE (n:{label} {{id: $id}}) "
                 f"ON CREATE SET {set_clauses} "
                 f"ON MATCH SET {set_clauses}")
        params = {"id": node_id}
        params.update(properties)
        self.session.run(query, params)

    def create_relationship(self, start_id, end_id, rel_type):
        self.session.run(
            f"MATCH (a {{id:$startId}}), (b {{id:$endId}}) MERGE (a)-[:{rel_type}]->(b)",
            {"startId": start_id, "endId": end_id}
        )

    def visit_ClassDef(self, node):
        class_name = node.name
        class_id = f"{self.root_name}:{class_name}"
        url = self.create_url(self.file_path, node.lineno)

        # Create or merge Class node with color
        self.merge_node("Class", class_id, {
            "name": class_name,
            "project": self.root_name,
            "file": self.file_path,
            "url": url,
            "color": NODE_COLORS["Class"]
        })

        # Link Root to Class
        self.create_relationship(self.root_id, class_id, "DECLARES")

        # Process methods
        self.current_class = class_id
        self.generic_visit(node)
        self.current_class = None

    def visit_FunctionDef(self, node):
        function_name = node.name
        url = self.create_url(self.file_path, node.lineno)
        if self.current_class:
            # It's a method
            function_id = f"{self.current_class}.{function_name}"
            self.merge_node("Method", function_id, {
                "name": function_name,
                "classId": self.current_class,
                "project": self.root_name,
                "file": self.file_path,
                "url": url,
                "color": NODE_COLORS["Method"]
            })
            # Link Class to Method
            self.create_relationship(self.current_class, function_id, "DECLARES")
        else:
            # It's a function
            function_id = f"{self.root_name}:{function_name}"
            self.merge_node("Function", function_id, {
                "name": function_name,
                "project": self.root_name,
                "file": self.file_path,
                "url": url,
                "color": NODE_COLORS["Function"]
            })
            # Link Root to Function
            self.create_relationship(self.root_id, function_id, "DECLARES")

        self.generic_visit(node)

def analyze_directory(dir_path, session, root_name, base_url):
    for root, dirs, files in os.walk(dir_path):
        for file in files:
            if file.endswith('.py') and not file.startswith('.'):
                file_path = os.path.join(root, file)
                analyze_file(file_path, session, root_name, base_url)

def main():
    root_name = os.getenv('ROOT_NAME') or 'UnknownRoot'
    base_url = os.getenv('BASE_URL') or 'http://localhost'
    project_root = '/app'  # Not directly used here, but can be utilized in URL creation

    driver = create_neo4j_driver()
    with driver.session() as session:
        # Clean up previous data and create constraints
        cleanup_previous_run_data(session, root_name)
        create_uniqueness_constraints(session)

        # Create root node with a unique id
        root_id = f"root:{root_name}"
        session.run(
            "MERGE (r:Root {id:$id}) "
            "ON CREATE SET r.name=$name, r.project=$project, r.color=$color "
            "ON MATCH SET r.name=$name, r.project=$project, r.color=$color",
            {"id": root_id, "name": root_name, "project": root_name, "color": NODE_COLORS["Root"]}
        )

        # Analyze the project directory
        analyze_directory('/app', session, root_name, base_url)

    driver.close()

if __name__ == "__main__":
    main()

File: js/analyzer.js
const fs = require("fs");
const path = require("path");
const babelParser = require("@babel/parser");
const traverse = require("@babel/traverse").default;
const neo4j = require("neo4j-driver");

const NodeColors = {
	Root: "orange",
	Class: "#4287f5",
	Function: "#42f54e",
	Method: "#42f54e",
	ReactComponent: "#f54242",
	ExternalService: "#f5f542",
	MongoCollection: "#f5a442",
};

async function createNeo4jDriver(uri, username, password) {
	const maxRetries = 100;
	let delay = 5000;

	for (let attempt = 0; attempt < maxRetries; attempt++) {
		try {
			const driver = neo4j.driver(uri, neo4j.auth.basic(username, password));
			const session = driver.session();
			await session.run("RETURN 1");
			await session.close();
			console.log("Connected to Neo4j");
			return driver;
		} catch (error) {
			console.log(
				`Attempt ${attempt + 1}: Neo4j connection failed, retrying in ${
					delay / 1000
				} seconds...`
			);
			await new Promise((resolve) => setTimeout(resolve, delay));
			delay *= 2;
		}
	}
	throw new Error("Failed to connect to Neo4j after several attempts");
}

async function createUniquenessConstraints(session) {
	await session.run(
		"CREATE CONSTRAINT IF NOT EXISTS FOR (r:Root) REQUIRE r.id IS UNIQUE"
	);
	await session.run(
		"CREATE CONSTRAINT IF NOT EXISTS FOR (c:Class) REQUIRE c.id IS UNIQUE"
	);
	await session.run(
		"CREATE CONSTRAINT IF NOT EXISTS FOR (f:Function) REQUIRE f.id IS UNIQUE"
	);
	await session.run(
		"CREATE CONSTRAINT IF NOT EXISTS FOR (m:Method) REQUIRE m.id IS UNIQUE"
	);
	await session.run(
		"CREATE CONSTRAINT IF NOT EXISTS FOR (rc:ReactComponent) REQUIRE rc.id IS UNIQUE"
	);
}

async function cleanupPreviousRunData(session, rootName) {
	await session.run("MATCH (n) WHERE n.project = $project DETACH DELETE n", {
		project: rootName,
	});
	console.log("Cleaned up data from previous run");
}

function createUrl(baseUrl, filePath, projectRoot, lineNumber) {
	const relativePath = path.relative(projectRoot, filePath);
	// Ensure baseUrl ends with '/'
	if (!baseUrl.endsWith("/")) {
		baseUrl += "/";
	}
	return `${baseUrl}${relativePath}#${lineNumber}`;
}

async function analyzeFile(
	filePath,
	session,
	rootName,
	baseUrl,
	projectRoot,
	rootId
) {
	const code = fs.readFileSync(filePath, "utf-8");
	let ast;
	try {
		ast = babelParser.parse(code, {
			sourceType: "module",
			plugins: [
				"jsx",
				"typescript",
				"classProperties",
				"decorators-legacy",
				"exportDefaultFrom",
				"exportNamespaceFrom",
				"dynamicImport",
				"objectRestSpread",
				"optionalChaining",
				"nullishCoalescingOperator",
			],
		});
	} catch (err) {
		console.error(`Failed to parse ${filePath}:`, err);
		return;
	}

	// We'll store queries and run them after traversal to ensure consistency.
	const queries = [];
	const processedNodes = new Set();

	function mergeNodeQuery(
		labels,
		id,
		properties,
		onCreateExtra = "",
		onMatchExtra = ""
	) {
		const labelString = Array.isArray(labels) ? labels.join(":") : labels;
		let setClause = Object.keys(properties)
			.map((key) => `n.${key} = $${key}`)
			.join(", ");
		let query = `MERGE (n:${labelString} {id: $id}) ON CREATE SET ${setClause}`;
		if (onCreateExtra) query += `, ${onCreateExtra}`;
		if (onMatchExtra) query += ` ON MATCH SET ${setClause}`;
		return { query, params: { id, ...properties } };
	}

	function createRelationshipQuery(sourceId, targetId, relType) {
		return {
			query: `MATCH (a {id: $sourceId}), (b {id: $targetId}) MERGE (a)-[:${relType}]->(b)`,
			params: { sourceId, targetId },
		};
	}

	// Helpers to create different node types and link them to Root
	function ensureClassNode(classId, className, filePath, line) {
		if (!processedNodes.has(classId)) {
			processedNodes.add(classId);
			queries.push(
				mergeNodeQuery("Class", classId, {
					name: className,
					project: rootName,
					file: filePath,
					color: NodeColors.Class,
					url: createUrl(baseUrl, filePath, projectRoot, line),
				})
			);
			// Link Root to Class
			queries.push(createRelationshipQuery(rootId, classId, "DECLARES"));
		}
	}

	function ensureFunctionNode(funcId, funcName, filePath, line) {
		if (!processedNodes.has(funcId)) {
			processedNodes.add(funcId);
			queries.push(
				mergeNodeQuery("Function", funcId, {
					name: funcName,
					project: rootName,
					file: filePath,
					color: NodeColors.Function,
					url: createUrl(baseUrl, filePath, projectRoot, line),
				})
			);
			// Link Root to Function
			queries.push(createRelationshipQuery(rootId, funcId, "DECLARES"));
		}
	}

	function ensureMethodNode(methodId, methodName, filePath, line) {
		if (!processedNodes.has(methodId)) {
			processedNodes.add(methodId);
			queries.push(
				mergeNodeQuery("Method", methodId, {
					name: methodName,
					project: rootName,
					file: filePath,
					color: NodeColors.Method,
					url: createUrl(baseUrl, filePath, projectRoot, line),
				})
			);
			// Also link Root to Method to keep everything accessible
			queries.push(createRelationshipQuery(rootId, methodId, "DECLARES"));
		}
	}

	function ensureReactComponentNode(
		componentId,
		componentName,
		filePath,
		line
	) {
		if (!processedNodes.has(componentId)) {
			processedNodes.add(componentId);
			queries.push(
				mergeNodeQuery("ReactComponent", componentId, {
					name: componentName,
					project: rootName,
					file: filePath,
					color: NodeColors.ReactComponent,
					url: createUrl(baseUrl, filePath, projectRoot, line),
				})
			);
			// Link Root to ReactComponent
			queries.push(createRelationshipQuery(rootId, componentId, "DECLARES"));
		}
	}

	function ensureCalleeNode(calleeId, calleeName) {
		// Default to Function if unknown
		if (!processedNodes.has(calleeId)) {
			processedNodes.add(calleeId);
			queries.push(
				mergeNodeQuery("Function", calleeId, {
					name: calleeName,
					project: rootName,
				})
			);
			// Link Root to it as well
			queries.push(createRelationshipQuery(rootId, calleeId, "DECLARES"));
		}
	}

	function createCallsRelationship(callerId, calleeId) {
		queries.push(createRelationshipQuery(callerId, calleeId, "CALLS"));
	}

	traverse(ast, {
		enter(path) {
			if (path.node.loc) {
				path.node.loc.file = filePath;
			}
		},
		ClassDeclaration(path) {
			const { node } = path;
			if (!node.id) return;
			const className = node.id.name;
			const classId = `${rootName}:${className}`;
			const line = node.loc.start.line;
			ensureClassNode(classId, className, filePath, line);

			// Also handle class methods inside the class body
			node.body.body.forEach((classMember) => {
				if (
					classMember.type === "ClassMethod" &&
					classMember.key.type === "Identifier"
				) {
					const methodName = classMember.key.name;
					const methodId = `${classId}.${methodName}`;
					const methodLine = classMember.loc.start.line;
					ensureMethodNode(methodId, methodName, filePath, methodLine);

					// Link Class to Method
					queries.push(createRelationshipQuery(classId, methodId, "DECLARES"));
				}
			});
		},
		FunctionDeclaration(path) {
			const { node } = path;
			if (!node.id) return;
			const functionName = node.id.name;
			const functionId = `${rootName}:${functionName}`;
			const line = node.loc.start.line;
			ensureFunctionNode(functionId, functionName, filePath, line);
		},
		ClassMethod(path) {
			// For class methods handled inside ClassDeclaration, but in case we want standalone handling:
			const { node } = path;
			if (
				!path.parentPath.parent ||
				path.parentPath.parent.type !== "ClassDeclaration"
			)
				return;
			const className = path.parentPath.parent.id.name;
			const classId = `${rootName}:${className}`;
			if (node.key.type === "Identifier") {
				const methodName = node.key.name;
				const methodId = `${classId}.${methodName}`;
				const line = node.loc.start.line;
				ensureMethodNode(methodId, methodName, filePath, line);
				queries.push(createRelationshipQuery(classId, methodId, "DECLARES"));
			}
		},
		CallExpression(path) {
			const { node } = path;
			let calleeName = "";
			let calleeId = "";
			const callee = node.callee;
			if (callee.type === "Identifier") {
				calleeName = callee.name;
				calleeId = `${rootName}:${calleeName}`;
			} else if (callee.type === "MemberExpression") {
				const objectName =
					callee.object && callee.object.type === "Identifier"
						? callee.object.name
						: "unknownObj";
				const propertyName =
					callee.property && callee.property.type === "Identifier"
						? callee.property.name
						: "unknownProp";
				calleeName = `${objectName}.${propertyName}`;
				calleeId = `${rootName}:${calleeName}`;
			} else {
				// Unsupported callee type, ignore
				return;
			}

			// Determine callerId
			const callerFunc = path.getFunctionParent();
			let callerId = "";
			if (callerFunc && callerFunc.node) {
				if (
					callerFunc.node.type === "FunctionDeclaration" &&
					callerFunc.node.id
				) {
					callerId = `${rootName}:${callerFunc.node.id.name}`;
				} else if (callerFunc.node.type === "ClassMethod") {
					const classNode = callerFunc.parentPath.parent;
					if (
						classNode &&
						classNode.type === "ClassDeclaration" &&
						classNode.id
					) {
						const className = classNode.id.name;
						const methodName = callerFunc.node.key.name;
						callerId = `${rootName}:${className}.${methodName}`;
					}
				}
			}
			// If we cannot determine callerId, skip relationship
			if (!callerId) {
				// Still ensure callee node exists if needed
				ensureCalleeNode(calleeId, calleeName);
				return;
			}

			// Ensure both caller and callee exist
			ensureCalleeNode(calleeId, calleeName);
			createCallsRelationship(callerId, calleeId);
		},
		JSXElement(path) {
			const { node } = path;
			const openingName = node.openingElement.name;
			if (openingName && openingName.type === "JSXIdentifier") {
				const componentName = openingName.name;
				const componentId = `${rootName}:${componentName}`;
				const line = node.loc.start.line;
				ensureReactComponentNode(componentId, componentName, filePath, line);
			}
		},
	});

	// After traversal, run all queries in a single transaction
	if (queries.length > 0) {
		const tx = session.beginTransaction();
		try {
			for (const q of queries) {
				await tx.run(q.query, q.params);
			}
			await tx.commit();
		} catch (err) {
			console.error(`Failed to execute queries for file ${filePath}:`, err);
			await tx.rollback();
		}
	}
}

async function analyzeDirectory(
	dir,
	session,
	rootName,
	baseUrl,
	projectRoot,
	rootId
) {
	const files = fs.readdirSync(dir);
	for (const file of files) {
		const filePath = path.join(dir, file);
		const stat = fs.lstatSync(filePath);
		if (stat.isDirectory()) {
			if (file === "node_modules" || file.startsWith(".")) {
				continue;
			}
			await analyzeDirectory(
				filePath,
				session,
				rootName,
				baseUrl,
				projectRoot,
				rootId
			);
		} else if (
			file.endsWith(".js") ||
			file.endsWith(".jsx") ||
			file.endsWith(".ts") ||
			file.endsWith(".tsx")
		) {
			await analyzeFile(
				filePath,
				session,
				rootName,
				baseUrl,
				projectRoot,
				rootId
			);
		}
	}
}

async function main() {
	const rootName = process.env.ROOT_NAME || "UnknownRoot";
	const baseUrl = process.env.BASE_URL || "http://localhost";
	const projectRoot = "/app";

	const neo4jUri = process.env.NEO4J_URI;
	const neo4jUser = process.env.NEO4J_USER;
	const neo4jPassword = process.env.NEO4J_PASSWORD;

	if (!neo4jUri || !neo4jUser || !neo4jPassword) {
		console.error(
			"Error: Missing required environment variables: NEO4J_URI, NEO4J_USER, NEO4J_PASSWORD"
		);
		process.exit(1);
	}

	const driver = await createNeo4jDriver(neo4jUri, neo4jUser, neo4jPassword);

	const session = driver.session();

	try {
		await cleanupPreviousRunData(session, rootName);
		await createUniquenessConstraints(session);

		const rootId = `root:${rootName}`;
		await session.run(
			"MERGE (r:Root {id:$id}) ON CREATE SET r.name=$name, r.project=$project, r.color=$color",
			{ id: rootId, name: rootName, project: rootName, color: NodeColors.Root }
		);

		await analyzeDirectory(
			projectRoot,
			session,
			rootName,
			baseUrl,
			projectRoot,
			rootId
		);
	} catch (error) {
		console.error("Error:", error);
	} finally {
		await session.close();
		await driver.close();
	}
}

main().catch((error) => {
	console.error("Main Error:", error);
});
